### PR TITLE
Protocol helpers

### PR DESCRIPTION
**Pull request by Joannis**

### CODE REVIEW

File: Sources/Async/File/File.swift [hunk 1]
```diff
@@ -1,6 +1,66 @@
 import Dispatch
 import Foundation
 
+/// An error converting types.
```
Comment (tanner0101) [inline Sources/Async/File/File.swift]: incorrect comments

File: Sources/Async/File/FileReader.swift [hunk 2]
```diff
@@ -5,8 +5,7 @@ import Foundation
 public protocol FileReader {
     /// Reads the file at the supplied path
     /// Supply a queue to complete the future on.
-    func read<S>(at path: String, into stream: S, chunkSize: Int)
-        where S: Async.InputStream, S.Input == UnsafeBufferPointer<UInt8>
+    func read(at path: String, chunkSize: Int) -> AnyOutputStream<UnsafeBufferPointer<UInt8>>
```
Comment (tanner0101) [inline Sources/Async/File/FileReader.swift]: What's wrong with providing the _into_ input stream? That allows us to keep the strong type info and makes implement a file reader simpler.

File: Sources/Async/Streams/ByteSerializerStream.swift [hunk 3]
```diff
@@ -0,0 +1,129 @@
+public protocol ByteSerializerStream: Async.Stream, ConnectionContext where Output == UnsafeBufferPointer<UInt8> {
+    var state: ByteSerializerStreamState<Self> { get }
+    
+    func serialize(_ input: Input) -> UnsafeBufferPointer<UInt8>
```
Comment (tanner0101) [inline Sources/Async/Streams/ByteSerializerStream.swift]: how exactly do you implement this function? are you required to manipulate the `state` variable manually or is it purely just serializing the message? what happens if the message is bigger than the buffer?

Comment (Joannis) [reply Sources/Async/Streams/ByteSerializerStream.swift]: The message is never bigger than the buffer. The buffer must be available asynchronously and is provided by the serializer. I'll document that

Comment (Joannis) [reply Sources/Async/Streams/ByteSerializerStream.swift]: Oh, I get it now. You mean multiple writes

File: Sources/Async/Streams/ByteParserStream.swift [hunk 4]
```diff
@@ -0,0 +1,181 @@
+public protocol ByteParserStream: Async.Stream, ConnectionContext where Input == UnsafeBufferPointer<UInt8> {
```
Comment (tanner0101) [inline Sources/Async/Streams/ByteParserStream.swift]: last we discussed, i thought we were going to have these extend TranslatingStream

File: Sources/Async/Streams/ByteParserStream.swift [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    var state: ByteParserStreamState<Self> { get }
+    
+    /// Continues parsing a partially parsed Output
+    func continueParsing(_ partial: Partial, from buffer: Input) throws -> ByteParserResult<Partial, Output>
+    
+    /// Starts parsing Output, resulting in either a partially or fully completed Output
+    func startParsing(from buffer: Input) throws -> ByteParserResult<Partial, Output>
+}
+
+public final class ByteParserStreamState<S: ByteParserStream> {
```
Comment (tanner0101) [inline Sources/Async/Streams/ByteParserStream.swift]: `<S> where S: ByteParserStream`

File: Sources/Async/Streams/ByteParserStream.swift [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    /// Stores partially parsed data
+    fileprivate var partiallyParsed: S.Partial?
+    
+    public init(worker: Worker) {
+        eventloop = worker.eventLoop
+        parsedInput = 0
+    }
+}
+
+public enum ByteParserResult<Partial, Output> {
```
Comment (tanner0101) [inline Sources/Async/Streams/ByteParserStream.swift]: this should probably just be generic on the byte parser stream as well. more future proof that way

File: Sources/Async/Streams/ByteParserStream.swift [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+extension ByteParserStream {
+    /// Parses the inputted byteBuffer to one or no output.
+    ///
+    /// If output has been achieved, passes it downstream and requests more data otherwise
+    ///
+    /// When output has been achieved, the remainder of the input buffer will be left unused until more output is requested.
+    public func translate(input: UnsafeBufferPointer<UInt8>) throws -> TranslatingStreamResult<Output> {
+        self.state.parsedInput = 0
+        
+        let state = try parseBytes(from: input, partial: self.state.partiallyParsed)
```
Comment (tanner0101) [inline Sources/Async/Streams/ByteParserStream.swift]: the input should be getting sliced here if not all of the bytes were consumed in the previous call

File: Sources/Async/Streams/ByteSerializerStream.swift [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    
+    /// Unsets all values, cleaning up the state
+    fileprivate func cancel() {
+        // clean up
+        backlog = []
+        consumedBacklog = 0
+    }
+    
+    /// Cleans up the backlog
+    fileprivate func cleanUpBacklog() {
```
Comment (tanner0101) [inline Sources/Async/Streams/ByteSerializerStream.swift]: i think backlog can be removed now

File: Sources/Async/Streams/PushStream.swift [hunk 9]
```diff
@@ -0,0 +1,94 @@
+public final class PushStream<Pushing>: Stream, ConnectionContext {
```
Comment (tanner0101) [inline Sources/Async/Streams/PushStream.swift:1]: maybe queue stream or backlog stream would be a better name

**Review by tanner0101 — COMMENTED** [review]
looking good.

File: Sources/Async/Streams/PushStream.swift (hunk 9 above)
Comment (Joannis) [reply Sources/Async/Streams/PushStream.swift:1]: Yes, it does keep a queue/backlog, but the gist of it (to me) is really conversion from push to pull

**Pull request merged**